src/adapter/philips_hue/mod.rs

```diff
@@ -0,0 +1,273 @@
+        PhilipsHueAdapter { name: "PhilipsHueAdapter".to_owned(),
+                       controller: controller,
+                       hubs: nupnp::query(),
+                     }
+    }
+}
+
+impl<T: Controller> ServiceAdapter for PhilipsHueAdapter<T> {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn start(&self) {
+        let mut id = 0;
+        let controller = self.controller.clone();
+        let hubs = self.hubs.clone();
+
+        thread::spawn(move || {
+            controller.send_event(
+                EventData::AdapterStart {
+                    name: "Philips Hue Service Adapter".to_owned()
+                }).unwrap();
+
+            debug!("nUPnP reported Philips Hue bridges: {:?}", hubs);
+            'pairing_check: for hub in hubs {
+                hub.is_available();
+                info!("Philips Hue bridge {} at {}", hub.id, hub.ip);
+
+                // Querying the settings fails when not paired with the bridge
+                if !hub.is_paired() {
+                    controller.send_event(
+                        EventData::AdapterInfo {
+                            info: format!("Push the pairing button on your Philips Hue bridge with ID {}",
+                            hub.id)
+                        }).unwrap();
+                    let pairing_result = hub.pair();
+                    if !pairing_result.contains("success") {
+                        info!("Giving up on pairing with Philips Hue bridge {}", hub.id);
+                        break 'pairing_check;
+                    }
+                    controller.send_event(
+                        EventData::AdapterInfo {
+                             info: format!("Pairing with Philips Hue bridge {} successfull",
+                             hub.id)
+                         }).unwrap();
+                }
+
+                // Extract and log some info
+                let setting = hub.get_settings();
+                let hs = HueSettings::new(&setting).unwrap(); // TODO: no unwrap
+                info!("Connected to Philips Hue bridge model {}, ID {}, software version {}, IP address {}",
+                    hs.config.modelid, hs.config.bridgeid, hs.config.swversion,
+                    hs.config.ipaddress);
+
+                let lights = hub.get_lights();
+                for light in lights {
+                    debug!("Spawning service for {:?}", light);
+                    id += 1;
+                    let service = PhilipsHueService::new(controller.clone(), id, light);
+                    let service_id = service.get_properties().id;
+                    service.start();
+                    controller.add_service(Box::new(service));
+                    controller.send_event(EventData::ServiceStart { id: service_id }).unwrap();
+                }
+            }
+        });
+    }
+
+    fn stop(&self) {
+        info!("Stopping Philips Hue adapter");
+    }
+
+}
+
+struct PhilipsHueService<T> {
```

Comment:
Done.

In fact they already have sensors. A "daylight sensor" is on every bridge. Doesn't seem to work for me. Might be connected to setting GPS coordinates and/or some external service. Also switches appear as sensors.